python/lsst/ctrl/mpexec/separablePipelineExecutor.py

```diff
@@ -84,9 +84,15 @@ class SeparablePipelineExecutor:
     butler : `lsst.daf.butler.Butler`
         A Butler whose ``collections`` and ``run`` attributes contain the input
         and output collections to use for processing.
+    extend_run : `bool`, optional
+        True if extending an existing run instead of creating a new empty run.
+        Quanta with complete outputs will be skipped. Partial outputs from
+        failed quanta will be removed if ``clobberOutputs`` is `True`,
+        otherwise an exception will be raised.
```

Comment:
I find this argument a bit confusing, and I think other users will too -- by design, `SeparablePipelineExecutor` does not do any collection management, and the concept of extending/not extending a run isn't well-defined in Prompt Processing. That could lead to a situation where the client doesn't know how this flag should be set.

I see you pass this flag to `SingleQuantumExecutor`. In the existing code, `PreExecInit` does its "extend" processing (i.e., tolerating existing datasets) if and only if `clobber_output` is set. Judging from the new docs added below, exactly the same reasoning applies to `SingleQuantumExecutor`.

Reply:
`PreExecInit` logic is a bit more complicated, but it does not care about `clobberOutputs` as it cannot clobber anything, it only knows about `extendRun`. When extending a run, we require that all in-memory initOutputs be identical or compatible with what is already stored in the output run. When `extendRun` is not set, `PreExecInit` verifies that none of the initOutputs is stored already. When `extendRun` is set, it tries to read existing initOutputs and compare them to in-memory data (at least for configs and packages). And it stores initOutpus that do not exist in either case.

`SingleQuantumExecutor` logic is simpler - if `extendRun` and `clobberOutputs` are both set then remove partial outputs. If `extendRun` is true then it skips quantum execution if full set of outputs exists (it does not remove full outputs). Previous version of the code could behave somewhat differently in case inconsistent options were given (e.g. `clobberOutputs` is True but `extendRun` is False).

From what you say it seems that you want to have an option to clobber full outputs from a successful quanta execution, I think this may require some collection management. @TallJimbo may provide more insight into that.

Reply:
> an option to clobber full outputs from a successful quanta execution

I don't think it needs collection management.  But I think it does mean `SingleQuantumExecutor` would require `extendRun=True` to be paired with _either_ `skipExisting=True` _or_ `clobberOutputs=True` (or both), and that would mean walking back some of the changes you just made.

I do think it'd still be fine for `pipetask` to have `--extend-run` imply `--skip-existing` because I don't think we need the prompt-processing clobber-everything behavior there, and I want to minimize how much that mode is used (I'd like to find a way to not need it in prompt processing, too, but I have not dug into the problem to see know if that might be feasible).

Reply:
Clobbering full outputs would be useful (at least, it matches my expectation of what "clobber" means), but I don't think it's the essential issue here.

My concern is that `extend_run` does not make sense as part of the `SeparablePipelineExecutor` API, because it suggests that there's `pipetask`-style collection management going on. As I noted above, when delegating to lower-level components that _do_ have a concept of extending runs, I've used `clobber_outputs` as a proxy (on the grounds that that's what you'd pass if you were doing any sort of retry). From @andy-slac's description, it sounds like `skip_existing` would actually be a better proxy when calling `SingleQuantumExecutor`, and in hindsight perhaps that's what I should have used for deciding how `PreExecInit` handles prior outputs as well.

At the very least, I think the calls to `PreExecInit` and `SingleQuantumExecutor` need to use the same strategy -- either the explicit `extend_run` flag added here, or a (matching?) proxy that expresses when their "extend-run" logic would apply to externally managed runs.

Reply:
I think you've got it right that `extendRun` isn't a meaningful concept at this low level, but that might just be a naming issue - at this level that value is still very useful as a way to express whether or not there is any chance of a collision between to-be-written datasets and existing datasets.  When there isn't we know we don't need to search for existing datasets, that can be fairly expensive.  Saying that we're not extending the run is sufficient (but not necessary way) to guarantee no collisions.

I can think of four states in `SingleQuantumExecutor` and `PreExecInit` that we have already tried to capture with these kinds of arguments:

1. Higher-level code guarantees that there is no chance of a collision, which may be because it's a new RUN or because it knows the data IDs that are present are disjoint with the new ones (though the latter is impossible for PreExecInit, since its outputs have empty data IDs).  This is the `extendRun=False` case.

2. There may be collisions, and the to-be-written datasets are always what we want, so clobber the old ones.  I am a bit skeptical of this state because it implies the new ones would not be the same as the old ones, which should only be possible if the outputs are not deterministic given the configs and software versions (which should be fixed, since collections only happen within a RUN), and that'd be a problem.  This is the `clobberOutputs=True`, `skipExisting=False` case.

3. There may be collisions, we expect the to-be-written ones to be equivalent to the old ones, and we want to retry any quanta that were not successful before, so clobber outputs from incomplete/failed quanta only.  This is the `clobberOutputs=True`, `skipExisting=True` case.

4. There may be collisions, we expect the to-be-written ones to be equivalent to the old ones, and we want to preserve those failures (and hold all downstream quanta).  This is the old `skipExisting=True` with `clobberOutputs=False`.

I think we need to keep 1, 3, and 4, but writing this up has reminded me why I did feel comfortable dropping (2) before, at least if there isn't a need for it in Prompt Processing.

Reply:
From the list above, `skipExisting` and `clobberOutputs` can exist in any combination, but is `extendRun` actually an extra dimension of that space? For example, is the combination `extend=True, clobber=False, skip=False` valid?

Reply:
Sorry for delay, some usual distractions from other projects. I also looked at how all these options are used in GraphBuilder and PreExecInit, and I'm ready to get back to SingleQuantumExecutor. I think I can make it work by just having `skipExisting` and `clobberOutputs` options (and no `extendRun`). The logic would be:
- if full outputs (metadata or other outputs) exist:
  - if `skipExisting` is True then quantum execution is skipped
  - otherwise if `clobberOutputs` is true they are deleted
  - otherwise an error generated
- else if partial inputs are found:
  - if `clobberOutputs` is true they are deleted
  - otherwise an error generated

One minor issue is that `skipExisting` is a "derivative" of the `skip_existing_in`. The latter is needed to pass it to `GraphBuilder`, so `SeparablePipelineExecutor` already defines it. In that sense, I think it would be better for `SingleQuantumExecutor` to accept `skip_existing_in` instead of `skip_existing` (and this is how it was before my last commit).

Reply:
I think that misses out on a potentially important optimization of not checking for existing outputs at all, but I don't remember what we're doing in that respect right now - if we don't have a branch for that possibility already, let's go with your plan and come back to this later.  If we do have a branch for that, I think we should find a way to keep it to avoid a performance regression.

Reply:
OK, I have re-implemented everything which has undone many of the changes on my previous commit. `SingleQuantumExecutor` interface is now unchanged from `main`, this means no changes to `SeparablePipelineExecutor`. I squashed it into one commit [7e0685b](https://github.com/lsst/ctrl_mpexec/pull/236/commits/7e0685bd718993c1f030b3eb8310defab80a63c8) to reduce noise. @TallJimbo, could you have another look.